### PR TITLE
Only use alphanumeric + some special characters in filenames

### PR DESCRIPTION
**Pull request by natri0**
As suggested by @ekrtcb in [a comment](https://github.com/ppy/osu/issues/20858#issuecomment-1290344195) to #20858, instead of having a list of disallowed characters we could only allow `0-9A-Za-z-_()[]` or a similar list.

The way the characters are filtered could probably be improved, though..

### CODE REVIEW

**Comment by AstroOrbis** [comment]
Couldn't we use a regex similar to [this?](https://regex101.com/r/Sgif2R/3)

**Comment by AstroOrbis** [comment]
> > Couldn't we use a regex similar to [this?](https://regex101.com/r/Sgif2R/3)
> 
> Is `0` dissallowed?

I believe \d means "any digit", so `0` should be allowed. Looks to be that way in testing too

**Comment by bdach** [comment]
Tend to agree with using a regex for this, but the one provided by @AstroOrbis is grossly incorrect, which is even proved by the link to the regex checker provided above.

**Comment by AstroOrbis** [comment]
> Tend to agree with using a regex for this, but the one provided by @AstroOrbis is grossly incorrect, which is even proved by the link to the regex checker provided above.
> 
> You'd want something like `[A-Za-z0-9\-_\(\)\[\]]+` to match this PR's current behaviour, I believe. (Note that the `\w` class permits a lot of _other_ extended latin characters. Which may be fine, may be not.)

I spent around 5 minutes on it, and trust me, I'm really bad at regex- it was just a suggestion to be worked on by someone more experienced than me.

(Ps: Thanks for the tip about `\w`)

**Comment by NullifiedJosh** [comment]
> > > Couldn't we use a regex similar to [this?](https://regex101.com/r/Sgif2R/3)
> > 
> > 
> > Is `0` dissallowed?
> 
> I believe \d means "any digit", so `0` should be allowed. Looks to be that way in testing too

Hmm, Maybe something wrong on my end.

**Comment by peppy** [comment]
> Tend to agree with using a regex for this

Curious on your reasoning for preferring regex. I don't really have an opinion each way, but generally avoid using regex for simple cases like this as it would be less-performant.

**Comment by bdach** [comment]
My reasoning is mostly brevity of the code, but I suppose it is personal preference. I'm not sure performance here is going to realistically matter here, although you're right that a regex is likely slower (maybe unless a static compiled instance is used).

**Comment by peppy** [comment]
Yeah, it's likely less of a concern for this specific pathway. I'm fine with either anyway.

**Comment by nullium21** [comment]
> My reasoning is mostly brevity of the code, but I suppose it is personal preference.

Cooked up [one that seems to work](https://regex101.com/r/e6t4OF/2); it matches the invalid characters one at a time so that we can do `regex.Replace(filename, "_")`. This way, the code would actually look nicer IMO, thanks.

Also, if performance does actually matter here, we can precompile the regex with `RegexOptions.Compiled` ¯\\\_(ツ)_/¯

**Comment by nullium21** [comment]
Apparently, `TestMetadata` ([`osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs:35`](https://github.com/nullium21/osu/blob/fix-filename-char-filtering/osu.Game.Tests/Visual/Editing/TestSceneEditorSaving.cs#L35)) fails with NullReferenceException after `SaveEditor();`... tried running it with a debugger and `EditorBeatmap` seems to be null. why? idk.

**Comment by bdach** [comment]
The tests are failing because the allowed set of characters does not contain the dot (`.`) character, and thereby breaking all files with extensions (because the dot, being the extension delimiter, was replaced with an underscore).

File: osu.Game/Extensions/ModelExtensions.cs [hunk 1]
```diff
@@ -137,20 +137,14 @@ private static bool matchesOnlineID(this IHasOnlineID<int>? instance, IHasOnline
             return instance.OnlineID.Equals(other.OnlineID);
         }
 
-        private static readonly char[] invalid_filename_characters = Path.GetInvalidFileNameChars()
-                                                                         // Backslash is added to avoid issues when exporting to zip.
-                                                                         // See SharpCompress filename normalisation https://github.com/adamhathcock/sharpcompress/blob/a1e7c0068db814c9aa78d86a94ccd1c761af74bd/src/SharpCompress/Writers/Zip/ZipWriter.cs#L143.
-                                                                         .Append('\\')
-                                                                         .ToArray();
+        private static Regex invalid_filename_chars = new(@"(?!$)[^A-Za-z0-9_()[\] -]", RegexOptions.Compiled);
```
Comment (bdach) [inline osu.Game/Extensions/ModelExtensions.cs:140]:
```suggestion
        private static readonly Regex invalid_filename_chars = new Regex(@"(?!$)[^A-Za-z0-9_()[\] -]", RegexOptions.Compiled);
```

We don't use target-typed `new` in this repository.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Extensions/ModelExtensions.cs (hunk 1 above)
Comment (AstroOrbis) [inline osu.Game/Extensions/ModelExtensions.cs:140]: I believe that `(?!$)[^A-Za-z0-9_()[\] -\\.]` includes the . character as well.

Comment (nullium21) [reply osu.Game/Extensions/ModelExtensions.cs]: Yeah, changed it. Should be careful to not make it a range ("` -.`" would match any character between space and `.`).

Comment (nullium21) [reply osu.Game/Extensions/ModelExtensions.cs]: Fixed.

**Comment by nullium21** [comment]
lol. note for myself: 're-request review' removes all previous requests.

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
@bdach I've tidied this up, leaving this to you to do a final check.

**Review by bdach — APPROVED** [review]
Seems ok

**Pull request merged**